// syncano-server/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._class = exports.data = exports.response = exports.logger = exports.socket = exports.event = exports.channel = exports.instance = exports.account = exports.users = undefined;

var _server = require('./server');

var _server2 = _interopRequireDefault(_server);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _ref = new _server2.default(),
    users = _ref.users,
    account = _ref.account,
    instance = _ref.instance,
    channel = _ref.channel,
    event = _ref.event,
    socket = _ref.socket,
    logger = _ref.logger,
    response = _ref.response,
    data = _ref.data,
    _class = _ref._class;

exports.users = users;
exports.account = account;
exports.instance = instance;
exports.channel = channel;
exports.event = event;
exports.socket = socket;
exports.logger = logger;
exports.response = response;
exports.data = data;
exports._class = _class;
exports.default = _server2.default;
