// syncano-server/lib/query-builder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _nodeFetch = require('node-fetch');

var _nodeFetch2 = _interopRequireDefault(_nodeFetch);

var _utils = require('./utils');

var _settings = require('./settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var QueryBuilder = function () {
  function QueryBuilder() {
    _classCallCheck(this, QueryBuilder);

    this.baseUrl = 'https://' + (0, _settings.getHost)();
  }

  _createClass(QueryBuilder, [{
    key: 'fetch',
    value: function fetch(url, options) {
      var headers = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

      var headersToSend = Object.assign({
        'content-type': 'application/json',
        'x-api-key': this.instance.token
      }, headers);

      return (0, _nodeFetch2.default)(url, _extends({
        headers: headersToSend
      }, options)).then(_utils.parseJSON).then(_utils.checkStatus);
    }
  }, {
    key: 'nonInstanceFetch',
    value: function nonInstanceFetch(url, options, headers) {
      return (0, _nodeFetch2.default)(url, _extends({
        headers: _extends({
          'content-type': 'application/json'
        }, headers)
      }, options)).then(_utils.parseJSON).then(_utils.checkStatus);
    }
  }, {
    key: 'withQuery',
    value: function withQuery(query) {
      this._query = Object.assign({}, this.query, query);

      return this;
    }
  }, {
    key: 'withRelationships',
    value: function withRelationships(relationships) {
      this._relationships = this.relationships.concat(relationships);

      return this;
    }
  }, {
    key: 'withMappedFields',
    value: function withMappedFields(fields) {
      this._mappedFields = Object.assign.apply(Object, [{}, this.mappedFields].concat(_toConsumableArray(fields)));

      return this;
    }
  }, {
    key: 'query',
    get: function get() {
      return this._query || {};
    }
  }, {
    key: 'relationships',
    get: function get() {
      return this._relationships || [];
    }
  }, {
    key: 'mappedFields',
    get: function get() {
      return this._mappedFields || [];
    }
  }]);

  return QueryBuilder;
}();

exports.default = QueryBuilder;
module.exports = exports['default'];
